File: chess/java/src/main/java/com/example/chess/game/Game.java
package com.example.chess.game;

import com.example.chess.exception.InvalidMoveException;
import com.example.chess.model.Board;
import com.example.chess.model.Color;
import com.example.chess.model.pieces.Piece;

import java.util.ArrayList;
import java.util.List;

/**
 * Application service for turn-based play.
 *
 * <p>Game enforces orchestration rules (whose turn, source has a piece, no own capture), then
 * delegates geometry to the piece polymorphically. Because chess is turn-based, this MVP needs no
 * concurrency control; a UI/API would serialize one move request at a time.
 */
public class Game {
    private final Board board;
    private final List<Piece> capturedPieces = new ArrayList<>();
    private Color currentTurn;

    public Game() {
        this(new Board(), Color.WHITE);
    }

    public Game(Board board, Color startingTurn) {
        this.board = board;
        this.currentTurn = startingTurn;
    }

    public void makeMove(String from, String to) {
        makeMove(Move.of(from, to));
    }

    public void makeMove(Move move) {
        Piece piece = board.getPiece(move.from())
                .orElseThrow(() -> new InvalidMoveException("No piece at " + move.from().toAlgebraic()));
        if (piece.getColor() != currentTurn) {
            throw new InvalidMoveException("It is " + currentTurn + "'s turn, not " + piece.getColor());
        }
        if (board.hasOwnPiece(move.to(), currentTurn)) {
            throw new InvalidMoveException("Cannot capture your own piece at " + move.to().toAlgebraic());
        }
        if (!piece.isValidMove(board, move.from(), move.to())) {
            throw new InvalidMoveException("Illegal " + piece.getClass().getSimpleName()
                    + " move: " + move.from().toAlgebraic() + " -> " + move.to().toAlgebraic());
        }
        board.getPiece(move.to()).ifPresent(capturedPieces::add);
        board.movePiece(move.from(), move.to());
        currentTurn = currentTurn.opposite();
    }

    public boolean isInCheck(Color color) {
        return board.isInCheck(color);
    }

    public Board getBoard() {
        return board;
    }

    public Color getCurrentTurn() {
        return currentTurn;
    }

    public List<Piece> getCapturedPieces() {
        return List.copyOf(capturedPieces);
    }
}

File: chess/python/chess/game.py
"""Turn-based game service: validates and applies moves."""

from __future__ import annotations

from dataclasses import dataclass

from .board import Board
from .exceptions import InvalidMoveError
from .models import Cell, Color
from .pieces import Piece


@dataclass(frozen=True)
class Move:
    """Immutable value object representing one requested move."""

    from_cell: Cell
    to_cell: Cell

    @classmethod
    def of(cls, from_square: str, to_square: str) -> "Move":
        return cls(Cell.from_algebraic(from_square), Cell.from_algebraic(to_square))


class Game:
    """Application service for turn-based play.

    Game enforces orchestration rules (whose turn, source has a piece, no own capture), then
    delegates geometry to the piece polymorphically. Chess is turn-based, so this MVP needs no
    concurrency control; a UI/API would serialize one move request at a time.
    """

    def __init__(self, board: Board | None = None, starting_turn: Color = Color.WHITE) -> None:
        self.board = board or Board()
        self.current_turn = starting_turn
        self.captured_pieces: list[Piece] = []

    def make_move(self, from_square: str | Move, to_square: str | None = None) -> None:
        move = from_square if isinstance(from_square, Move) else Move.of(from_square, to_square or "")
        piece = self.board.get_piece(move.from_cell)
        if piece is None:
            raise InvalidMoveError(f"No piece at {move.from_cell.to_algebraic()}")
        if piece.color is not self.current_turn:
            raise InvalidMoveError(f"It is {self.current_turn.value}'s turn, not {piece.color.value}")
        if self.board.has_own_piece(move.to_cell, self.current_turn):
            raise InvalidMoveError(f"Cannot capture your own piece at {move.to_cell.to_algebraic()}")
        if not piece.is_valid_move(self.board, move.from_cell, move.to_cell):
            raise InvalidMoveError(
                f"Illegal {piece.__class__.__name__} move: "
                f"{move.from_cell.to_algebraic()} -> {move.to_cell.to_algebraic()}"
            )
        captured = self.board.get_piece(move.to_cell)
        if captured is not None:
            self.captured_pieces.append(captured)
        self.board.move_piece(move.from_cell, move.to_cell)
        self.current_turn = self.current_turn.opposite

    def is_in_check(self, color: Color) -> bool:
        return self.board.is_in_check(color)
